### PR TITLE
[Merged by Bors] - Log of state root

### LINKED ISSUE #1547: don't panic when block from hare result not found
# Overview / Motivation
currently node panics when a block found in hare results is not in nodes database 
# The Task
remove the panic, replace with error, send nodes view in created block and not hare result set. 

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who haven't introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### LINKED ISSUE #1591: Ensure we have logs of state root
- [x] Log global state root whenever it's updated
- [x] Log layer hash (concatenation of sorted block ids + prev. layer hash) every time pbase passes it
- [x] Ensure automation tests validate that all nodes have the same mesh at the end of the test
- [x] Fix #1547

### PR DESCRIPTION
**Pull request by noamnelke**
## Motivation
Closes #1591 
Closes #1547 

## Changes
- [x] Log every state root history change.
- [x] Log "layer hash" on every pbase advance.
- [x] Validate state roots and layer hashes are same for all nodes at end of tests.
- [x] Resolves #1547 (ignore missing blocks in hare results, assuming they will be received later).

## Test Plan
- [x] Add assertions to existing tests.

### CODE REVIEW

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

File: mesh/mesh.go [hunk 1]
```diff
@@ -181,10 +187,50 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
...
 
+func (m *Mesh) logStateRoot(layerId types.LayerID) {
+	m.Event().Info("end of layer state root",
+		log.LayerId(layerId.Uint64()),
+		log.String("state_root", util.Bytes2Hex(m.TxProcessor.GetStateRoot().Bytes())),
+	)
+}
+
+func (m *Mesh) setLayerHash(layerId types.LayerID) {
+	validBlockIds, err := m.ContextuallyValidBlock(layerId)
```
Comment (antonlerner) [inline mesh/mesh.go]: this is code duplication from `ExtractUniqueOrderedTransactions` , try to refactor

File: mesh/mesh.go [hunk 2]
```diff
@@ -181,10 +187,50 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
...
+	for _, id := range sortedBlockIds {
+		layerHash.Write(id.ToBytes())
+	}
+	m.layerHash = layerHash.Sum(nil)
+	m.Event().Info("new layer hash",
+		log.LayerId(layerId.Uint64()),
+		log.String("layer_hash", util.Bytes2Hex(m.layerHash)))
+}
+
+func (m *Mesh) persistLayerHash() {
```
Comment (antonlerner) [inline mesh/mesh.go:215]: does this persist only the last layer hash? if so make sure it is understandable from the func name, however I think we should keep a key value between all layers and respeceted hashes

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

File: mesh/mesh.go [hunk 3]
```diff
@@ -176,42 +181,55 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
...
 
 	// Deterministically sort valid blocks
-	validBlocks = types.SortBlocks(validBlocks)
-
-	// `layerHash` is the sha256 sum of sorted layer block IDs
-	layerHash := sha256.New()
-	for _, b := range validBlocks {
-		layerHash.Write(b.Id().AsHash32().Bytes())
-	}
+	types.SortBlocks(validBlocks)
```
Comment (almogdepaz) [inline mesh/mesh.go:227]: why not use CalcBlocksHash32 ?

Comment (noamnelke) [reply mesh/mesh.go:227]: I need to have to blocks sorted for later in the method

File: common/types/hashes.go [hunk 4]
```diff
@@ -108,17 +108,22 @@ func CalcBlocksHash12(view []BlockID) (Hash12, error) {
...
-		return Hash32{}, err
+	SortBlockIds(sortedView)
+	return CalcBlockHash32Presorted(sortedView, additionalBytes)
+}
+
+func CalcBlockHash32Presorted(sortedView []BlockID, additionalBytes []byte) Hash32 {
+	hash := sha256.New()
+	hash.Write(additionalBytes)
+	for _, id := range sortedView {
+		hash.Write(id.ToBytes())
```
Comment (almogdepaz) [inline common/types/hashes.go]: hash.Write errors are not handled

File: mesh/meshdb.go [hunk 5]
```diff
@@ -223,23 +208,22 @@ func (m *MeshDB) ForBlockInView(view map[types.BlockID]struct{}, layer types.Lay
...
 		return nil, fmt.Errorf("error getting layer %v from database %v", index, err)
 	}
 
-	if len(ids) == 0 {
+	if len(idsBytes) == 0 {
 		return nil, fmt.Errorf("no ids for layer %v in database ", index)
 	}
 
-	idSet, err := types.BytesToBlockIds(ids)
+	blockIds, err := types.BytesToBlockIds(idsBytes)
```
Comment (almogdepaz) [inline mesh/meshdb.go:221]: "idSet" name was on purpose to signify that there are no repetitions, why change it ?

File: mesh/meshdb.go [hunk 6]
```diff
@@ -223,23 +208,22 @@ func (m *MeshDB) ForBlockInView(view map[types.BlockID]struct{}, layer types.Lay
 	return nil
 }
 
-func (m *MeshDB) layerBlockIds(index types.LayerID) ([]types.BlockID, error) {
-
-	ids, err := m.layers.Get(index.ToBytes())
+func (m *MeshDB) LayerBlockIds(index types.LayerID) ([]types.BlockID, error) {
```
Comment (almogdepaz) [inline mesh/meshdb.go:211]: why did you change this to public? looks like no one is using it outside this pkg

File: mesh/mesh.go (hunk 3 above)
Comment (almogdepaz) [reply mesh/mesh.go]: yes but CalcBlocksHash32 also sorts the blocks (it has to)

Comment (almogdepaz) [reply mesh/mesh.go:227]: cool i get that your trying to save on the sort. but i would consider sorting twice and making CalcBlockHash32Presorted private just to have leaner code and maybe prevent someone using it wrong in the future.
sorting the 200 ids twice is negligible in my opinion

Comment (almogdepaz) [reply mesh/mesh.go]: also CalcBlockHash32Presorted sounded to me like the method expects a presorted slice, might just be me though ...

File: common/types/hashes.go (hunk 4 above)
Comment (noamnelke) [reply common/types/hashes.go]: It never returns an error: https://golang.org/pkg/hash/#Hash

File: mesh/mesh.go (hunk 3 above)
Comment (noamnelke) [reply mesh/mesh.go]: It expects a pre sorted slice. What makes you think otherwise?

Comment (noamnelke) [reply mesh/mesh.go:227]: Maybe you’re right. It felt more “correct” to me this way. 

I’ll take another look at it tomorrow.

File: mesh/meshdb.go (hunk 6 above)
Comment (noamnelke) [reply mesh/meshdb.go:211]: The `ninjaTortoise` uses it:
image

Note that I didn't "change this to public". I just replaced a wrapper function that didn't seem to do anything with the actual function that the wrapper called. This is the deleted function:
https://github.com/spacemeshos/go-spacemesh/blob/66e9a7ead2d9b2f32e61cda2aad8b4ed5f4aedff/mesh/meshdb.go#L165-L178

File: mesh/meshdb.go (hunk 5 above)
Comment (noamnelke) [reply mesh/meshdb.go:221]: `Set` was confusing to me. I thought it signified a set data structure (e.g. implemented as a map), so I was surprised to see it was a slice. Maybe I'll change it to `uniqueIds` if you think it's important to signify it here. Maybe it would be even better to change the method name to `UniqueLayerBlockIds` if you think it's important to signify that, since an internal variable name isn't very visible.

I personally would expect `LayerBlockIds` not to return duplicates even without spelling it out, but I don't think any of this is critical. Just let me know what you think is best and I'll do it (including reverting to the original).

File: mesh/mesh.go (hunk 3 above)
Comment (noamnelke) [reply mesh/mesh.go:227]: I'm not sure about this.

This is copying and sorting 6.4 KB using a rather expensive comparison function. Not a big deal, but if we can easily avoid it, why not?

I'm actually considering doing the opposite... It looks like all other usages of `CalcBlocksHash32()` don't need the copy of the slice, they pass a slice that's only used for this method. So we could remove `CalcBlocksHash32()` and then just call something like `CalcBlockHash32Presorted(SortBlockIds(ids))`.

What do you think?

File: mesh/meshdb.go (hunk 6 above)
Comment (almogdepaz) [reply mesh/meshdb.go:211]: got it, very strange, i wonder how that happened...
maybe we had some issue with the slices returned from xdr ? (maybe just a bad merge)
anyway keep your fix if we run in to issues ill remember this change

File: mesh/meshdb.go (hunk 5 above)
Comment (almogdepaz) [reply mesh/meshdb.go:221]: i dont think its good practice to introduce unnecessary changes just because it makes sense to you, i can understand a typo fix but specific naming fixes can cause variation in convention,
also makes it harder to keep up with changes (more diffs in your pr), this is not to say that we don't need to refactor our code and maybe provide better naming.
do you know if this convention exists in other locations ? what would prevent me from "fixing" this and changing it back because it makes more sense to me ?

File: mesh/mesh.go (hunk 3 above)
Comment (almogdepaz) [reply mesh/mesh.go:227]: i usually prefer code that is less prone to bad use, also sorting an already sorted array is very fast, and the compare compare bytes slices doesn't seem expensive at all

File: mesh/meshdb.go (hunk 5 above)
Comment (noamnelke) [reply mesh/meshdb.go:221]: I agree. I made this change because I thought it was a leftover from changing the data structure (I thought that's what `Set` meant). That's why I suggested to revert the change if you feel that it's a better description of the variable (this answers your last question - I'm fine with you changing it back and actually suggested to do it myself).

I wasn't kidding - just let me know what you think is best here and I'll do it. I didn't ask for your review out of politeness - I really want your opinion.

File: mesh/mesh.go (hunk 3 above)
Comment (noamnelke) [reply mesh/mesh.go:227]: `CalcBlockHash32Presorted()` is prone to bad use? I disagree on that.

File: mesh/meshdb.go (hunk 5 above)
Comment (almogdepaz) [reply mesh/meshdb.go:221]: the specific change is really not that important and the new names are fine , just wanted to point out the problem in the process.

**Review by antonlerner — APPROVED** [review]

**Comment by noamnelke** [comment]
Bors merge
